File: src/engine.c
/*
 * The Tiechu engine for IBus
 *
 * Copyright (c) 2020-2022 Duy Thai,duythaia3@gmail.com
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301, USA.
 */

#include "ZTelexTable2.h"
#include "engine.h"
#include "CandidateArray.h"
#include "Logger.h"
#include "CandidateTable.h"
#include "HanvietTable.h"
#include "TelexEditor.h"
#include "Configuration.h"
#include "PropertyList.h"
#include "EnglishMode.h"
#include "VietnameseMode.h"
#include "HanvietMode.h"
#include "TeochewMode.h"
#include "ChineseMode.h"
#include "Database.h"
#include "Helper.h"
#include "Preedit.h"
#include "LTelexTable.h"
#include "ETelexTable.h"
#include "InterchangeMode.h"


/* functions prototype */
static void ibus_tiechu_engine_class_init(IBusTiechuEngineClass *klass);

static void ibus_tiechu_engine_init(IBusTiechuEngine *tiechu);

static void ibus_tiechu_engine_destroy(IBusTiechuEngine *tiechu);

static gboolean ibus_tiechu_engine_process_key_event(IBusEngine *engine, guint keyval, guint keycode, guint modifiers);

static void ibus_tiechu_engine_focus_in(IBusEngine *engine);

static void ibus_tiechu_engine_focus_out(IBusEngine *engine);

//static void ibus_tiechu_engine_reset(IBusEngine *engine);
static void ibus_tiechu_engine_enable(IBusEngine *engine);

static void ibus_tiechu_engine_disable(IBusEngine *engine);

static void ibus_engine_set_cursor_location(IBusEngine *engine, gint x, gint y, gint w, gint h);

static void ibus_tiechu_engine_set_capabilities(IBusEngine *engine, guint caps);

static void ibus_tiechu_engine_page_up(IBusEngine *engine);

static void ibus_tiechu_engine_page_down(IBusEngine *engine);

static void ibus_tiechu_engine_cursor_up(IBusEngine *engine);

static void ibus_tiechu_engine_cursor_down(IBusEngine *engine);

static void ibus_tiechu_engine_candidate_clicked(IBusEngine *engine, guint index, guint button, guint state);

static void ibus_tiechu_property_activate(IBusEngine *engine, const gchar *prop_name, gint prop_state);

static void ibus_tiechu_engine_property_show(IBusEngine *engine, const gchar *prop_name);

static void ibus_tiechu_engine_property_hide(IBusEngine *engine, const gchar *prop_name);

static void ibus_tiechu_engine_commit_string(IBusTiechuEngine *tiechu, const gchar *string);

//
static void property_activated(IBusEngine *engine, const gchar *property_ten, guint property_status);

G_DEFINE_TYPE (IBusTiechuEngine, ibus_tiechu_engine, IBUS_TYPE_ENGINE)

//

static void ibus_tiechu_engine_class_init(IBusTiechuEngineClass *klass){
    logger_log("engine class init...\n");
    IBusObjectClass *ibus_object_class=IBUS_OBJECT_CLASS (klass);
    IBusEngineClass *engine_class=IBUS_ENGINE_CLASS (klass);

    //
    ibus_object_class->destroy=(IBusObjectDestroyFunc) ibus_tiechu_engine_destroy;
    engine_class->process_key_event=ibus_tiechu_engine_process_key_event;
    engine_class->focus_out=ibus_tiechu_engine_focus_out;
    engine_class->focus_in=ibus_tiechu_engine_focus_in;
    engine_class->property_activate=property_activated;

    //
    engine_class->page_up=ibus_tiechu_engine_page_up;
    engine_class->page_down=ibus_tiechu_engine_page_down;

    //
    //engine_class->cursor_up = ibus_tiechu_engine_cursor_up;
    //engine_class->cursor_down = ibus_tiechu_engine_cursor_down;

    //
    engine_class->candidate_clicked=ibus_tiechu_engine_candidate_clicked;
}

void initialize(){
    logger_log("Tiechu engine init...\n");

    //load configuration
    configuration_load();

    //
    database_open();

    // create property list, and show it
    property_list_init();
    property_list_load();

    // initialize telex-related objects
    z_telex_table_2_init();
    ltelex_table_init();
    etelex_table_init();
    preedit_init();

    // initialize lookup table
    candidate_table_init();

    candidate_array_init();
    hanviet_table_init();
}

void uninitialize(){
    logger_log("Tiechu engine destroy...\n");

    // destroy property list
    property_list_destroy();

    //telex
    preedit_destroy();
    z_telex_table_2_destroy();
    ltelex_table_destroy();
    etelex_table_destroy();

    candidate_array_destroy();
    hanviet_table_destroy();


    // destroy lookup table
    candidate_table_destroy();

    //
    database_close();
}

static void ibus_tiechu_engine_init(IBusTiechuEngine *tiechu){
    initialize();
}

static void ibus_tiechu_engine_destroy(IBusTiechuEngine *tiechu){
    uninitialize();
    ((IBusObjectClass *) ibus_tiechu_engine_parent_class)->destroy((IBusObject *) tiechu);
}

static void ibus_tiechu_engine_focus_out(IBusEngine *engine){
    helper_clear_preedit((IBusTiechuEngine *) engine);
}

static void ibus_tiechu_engine_focus_in(IBusEngine *engine){
    logger_log("Tiechu received focus_in event");
    logger_backup();
    property_list_register((IBusTiechuEngine *) engine);
}

static void ibus_tiechu_engine_page_up(IBusEngine *engine){
    logger_log("ibus_tiechu_engine_page_up");
    if(candidate_table_showing()){
        candidate_table_page_up((IBusTiechuEngine *) engine);
    }
}

static void ibus_tiechu_engine_page_down(IBusEngine *engine){
    logger_log("ibus_tiechu_engine_page_down");
    if(candidate_table_showing()){
        candidate_table_page_down((IBusTiechuEngine *) engine);
    }
}

static void ibus_tiechu_engine_cursor_up(IBusEngine *engine){
    logger_log("ibus_tiechu_engine_cursor_up");
    if(candidate_table_showing()){
        candidate_table_cursor_up((IBusTiechuEngine *) engine);
    }
}

static void ibus_tiechu_engine_cursor_down(IBusEngine *engine){
    logger_log("ibus_tiechu_engine_cursor_down");
    if(candidate_table_showing()){
        candidate_table_cursor_down((IBusTiechuEngine *) engine);
    }
}

// processing when a candidate is selected using mouse
static void ibus_tiechu_engine_candidate_clicked(IBusEngine *engine, guint index, guint button, guint state){
    helper_commit_candidate_in_page((IBusTiechuEngine *) engine, index);
}

static void property_activated(IBusEngine *engine, const gchar *property_ten, guint property_status){
    logger_log("Property %s changed to %d", property_ten, property_status);
    if(g_strcmp0(property_ten, TVIETNAMESE_MODE_VALUE) == 0){
        if(property_status==PROP_STATE_CHECKED){
            configuration_set_selected_mode(TVIETNAMESE_MODE);
            property_list_update_symbol((IBusTiechuEngine *) engine);
            configuration_save();
        }
    }else if(g_strcmp0(property_ten, THANVIET_MODE_VALUE) == 0){
        if(property_status==PROP_STATE_CHECKED){
            configuration_set_selected_mode(THANVIET_MODE);
            property_list_update_symbol((IBusTiechuEngine *) engine);
            configuration_save();
        }
    }else if(g_strcmp0(property_ten, TTEOCHEW_MODE_VALUE) == 0){
        if(property_status==PROP_STATE_CHECKED){
            configuration_set_selected_mode(TTEOCHEW_MODE);
            property_list_update_symbol((IBusTiechuEngine *) engine);
            configuration_save();
        }
    }else if(g_strcmp0(property_ten, TCHINESE_MODE_VALUE) == 0){
        if(property_status==PROP_STATE_CHECKED){
            configuration_set_selected_mode(TCHINESE_MODE);
            property_list_update_symbol((IBusTiechuEngine *) engine);
            configuration_save();
        }
    }else if(g_strcmp0(property_ten, TENGLISH_MODE_VALUE) == 0){
        if(property_status==PROP_STATE_CHECKED){
            configuration_set_selected_mode(TENGLISH_MODE);
            property_list_update_symbol((IBusTiechuEngine *) engine);
            configuration_save();
        }
    }else if(g_strcmp0(property_ten, TREMEMBER_MODE_VALUE) == 0){
        configuration_set_remembering_mode(property_status);
        configuration_save();
    }else if(g_strcmp0(property_ten, TAUTO_SHOW_CANDIDATE_TABLE_VALUE) == 0){
        configuration_set_auto_show_candidate_table(property_status);
        configuration_save();
    }
}

static gboolean deliver_key(IBusTiechuEngine *tiechu, guint keyval, guint keycode, guint modifiers){
    //
    gint current_mode=configuration_get_current_mode();
    switch(current_mode){
        case TVIETNAMESE_MODE:
            return vietnamese_mode_process_key_event(tiechu, keyval, keycode, modifiers);
        case TTEOCHEW_MODE:
            return teochew_mode_process_key_event(tiechu, keyval, keycode, modifiers);
        case THANVIET_MODE:
            return hanviet_mode_process_key_event(tiechu, keyval, keycode, modifiers);
        case TCHINESE_MODE:
            return chinese_mode_process_key_event(tiechu, keyval, keycode, modifiers);
        case TENGLISH_MODE:
            return english_mode_process_key_event(tiechu, keyval, keycode, modifiers);
        default:
            return FALSE;
    }
}

static gboolean manage_mode(IBusTiechuEngine *tiechu, guint keyval, guint keycode, guint modifiers){
    gint current_mode=configuration_get_current_mode(); //current mode
    guint mode_vietnamese_key=(guint) configuration_get_vietnamese_mode_key();
    guint teochew_mode_key=(guint) configuration_get_teochew_mode_key();
    guint hanviet_mode_key=(guint) configuration_get_hanviet_mode_key();
    guint chinese_mode_key=(guint) configuration_get_chinese_mode_key();
    guint english_mode_key=(guint) configuration_get_english_mode_key();

    // vietnamese mode
    if((helper_is_control_pressed(modifiers))&&(keyval==mode_vietnamese_key)){
        if(current_mode == TVIETNAMESE_MODE){
            return TRUE;
        }
        configuration_set_selected_mode(TVIETNAMESE_MODE);
        property_list_update(tiechu);
        configuration_save();
        //helper_clear_preedit(tiechu);
        helper_commit_preedit(tiechu);
        return TRUE;
    }

    // Han Viet mode
    if((helper_is_control_pressed(modifiers))&&(keyval==hanviet_mode_key)){
        if(current_mode == THANVIET_MODE){
            return TRUE;
        }
        configuration_set_selected_mode(THANVIET_MODE);
        property_list_update(tiechu);
        configuration_save();
        //helper_clear_preedit(tiechu);
        helper_commit_preedit(tiechu);
        return TRUE;
    }

    // teochew mode
    if((helper_is_control_pressed(modifiers))&&(keyval==teochew_mode_key)){
        if(current_mode == TTEOCHEW_MODE){
            return TRUE;
        }
        configuration_set_selected_mode(TTEOCHEW_MODE);
        property_list_update(tiechu);
        configuration_save();
        //helper_clear_preedit(tiechu);
        helper_commit_preedit(tiechu);
        return TRUE;
    }

    // chinese mode
    if((helper_is_control_pressed(modifiers))&&(keyval==chinese_mode_key)){
        if(current_mode == TCHINESE_MODE){
            return TRUE;
        }
        configuration_set_selected_mode(TCHINESE_MODE);
        property_list_update(tiechu);
        configuration_save();
        //helper_clear_preedit(tiechu);
        helper_commit_preedit(tiechu);
        return TRUE;
    }

    // english mode
    if((helper_is_control_pressed(modifiers))&&(keyval==english_mode_key)){
        if(current_mode == TENGLISH_MODE){
            return TRUE;
        }
        configuration_set_selected_mode(TENGLISH_MODE);
        property_list_update(tiechu);
        configuration_save();
        //helper_clear_preedit(tiechu);
        helper_commit_preedit(tiechu);
        return TRUE;
    }

    //
    if(keyval==IBUS_KEY_Control_L||keyval==IBUS_KEY_Control_R){
        interchange_mode_save_begin_interchange_mode_time();
        return FALSE;
    }

    //
    if(keyval==IBUS_KEY_Shift_L || keyval==IBUS_KEY_Shift_R){
        interchange_mode_save_begin_interchange_mode_time();
        return FALSE;
    }

    //
    if(interchange_mode_is_in_interchange_mode()){
        interchange_mode_turn_shift_pressed_off();
        logger_log("Leave interchange mode");
    }
    return deliver_key(tiechu, keyval, keycode, modifiers);
}


static gboolean ibus_tiechu_engine_process_key_event(IBusEngine *engine,guint keyval,guint keycode,guint modifiers){
    //IBusText *text;
    IBusTiechuEngine *tiechu=(IBusTiechuEngine *) engine;

    //
    if(interchange_mode_is_in_interchange_mode()){
        if(helper_is_event_released(modifiers)){
            return FALSE;
        }
        return manage_mode(tiechu, keyval, keycode, modifiers);
    }

    //
    if(keyval==IBUS_KEY_Shift_L || keyval==IBUS_KEY_Shift_R){
        if(helper_is_event_released(modifiers)){
            if(!interchange_mode_is_shift_pressed_on()){
                return FALSE;
            }
            logger_log("Enter interchange mode");
            interchange_mode_save_begin_interchange_mode_time();
            return FALSE;
        }else{
            logger_log("Left Shift pressed");
            interchange_mode_turn_shift_pressed_on();
            return FALSE;
        }
    }

    //
    if(interchange_mode_is_in_interchange_mode()){
        logger_log("Leave interchange mode");
        interchange_mode_turn_shift_pressed_off();
    }

    //
    if (!helper_is_escape_pressed(keyval)) {
        return deliver_key(tiechu, keyval, keycode, modifiers);
    } else {
        return FALSE;
    }

}

File: engine/engine.py
# vim:set et sts=4 sw=4:
#
# ibus-tmpl - The Input Bus template project
#
# Copyright (c) 2007-2014 Peng Huang <shawn.p.huang@gmail.com>
#
# This program is free software; you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation; either version 2, or (at your option)
# any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.

# for python2
from __future__ import print_function

import enchant

from gi.repository import GLib
from gi.repository import IBus
from gi.repository import Pango

keysyms = IBus

class EngineEnchant(IBus.Engine):
    __gtype_name__ = 'EngineEnchant'
    __dict = enchant.Dict("en")

    def __init__(self):
        super(EngineEnchant, self).__init__()
        self.__is_invalidate = False
        self.__preedit_string = ""
        self.__lookup_table = IBus.LookupTable.new(10, 0, True, True)
        self.__prop_list = IBus.PropList()
        self.__prop_list.append(IBus.Property(key="test", icon="ibus-local"))
        print("Create EngineEnchant OK")

    def do_process_key_event(self, keyval, keycode, state):
        print("process_key_event(%04x, %04x, %04x)" % (keyval, keycode, state))
        # ignore key release events
        is_press = ((state & IBus.ModifierType.RELEASE_MASK) == 0)
        if not is_press:
            return False

        if self.__preedit_string:
            if keyval == keysyms.Return:
                self.__commit_string(self.__preedit_string)
                return True
            elif keyval == keysyms.Escape:
                self.__preedit_string = ""
                self.__update()
                return True
            elif keyval == keysyms.BackSpace:
                self.__preedit_string = self.__preedit_string[:-1]
                self.__invalidate()
                return True
            elif keyval == keysyms.space:
                if self.__lookup_table.get_number_of_candidates() > 0:
                    self.__commit_string(self.__lookup_table.get_current_candidate().text)
                else:
                    self.__commit_string(self.__preedit_string)
                return False
            elif keyval >= 49 and keyval <= 57:
                #keyval >= keysyms._1 and keyval <= keysyms._9
                index = keyval - keysyms._1
                candidates = self.__lookup_table.get_canidates_in_current_page()
                if index >= len(candidates):
                    return False
                candidate = candidates[index].text
                self.__commit_string(candidate)
                return True
            elif keyval == keysyms.Page_Up or keyval == keysyms.KP_Page_Up:
                self.page_up()
                return True
            elif keyval == keysyms.Page_Down or keyval == keysyms.KP_Page_Down:
                self.page_down()
                return True
            elif keyval == keysyms.Up:
                self.cursor_up()
                return True
            elif keyval == keysyms.Down:
                self.cursor_down()
                return True
            elif keyval == keysyms.Left or keyval == keysyms.Right:
                return True
        if keyval in range(keysyms.a, keysyms.z + 1) or \
            keyval in range(keysyms.A, keysyms.Z + 1):
            if state & (IBus.ModifierType.CONTROL_MASK | IBus.ModifierType.MOD1_MASK) == 0:
                self.__preedit_string += chr(keyval)
                self.__invalidate()
                return True
        else:
            if keyval < 128 and self.__preedit_string:
                self.__commit_string(self.__preedit_string)

        return False

    def __invalidate(self):
        if self.__is_invalidate:
            return
        self.__is_invalidate = True
        GLib.idle_add(self.__update)


    def do_page_up(self):
        if self.__lookup_table.page_up():
            self.page_up_lookup_table()
            return True
        return False

    def do_page_down(self):
        if self.__lookup_table.page_down():
            self.page_down_lookup_table()
            return True
        return False

    def do_cursor_up(self):
        if self.__lookup_table.cursor_up():
            self.cursor_up_lookup_table()
            return True
        return False

    def do_cursor_down(self):
        if self.__lookup_table.cursor_down():
            self.cursor_down_lookup_table()
            return True
        return False

    def __commit_string(self, text):
        self.commit_text(IBus.Text.new_from_string(text))
        self.__preedit_string = ""
        self.__update()

    def __update(self):
        preedit_len = len(self.__preedit_string)
        attrs = IBus.AttrList()
        self.__lookup_table.clear()
        if preedit_len > 0:
            if not self.__dict.check(self.__preedit_string):
                attrs.append(IBus.Attribute.new(IBus.AttrType.FOREGROUND,
                        0xff0000, 0, preedit_len))
                for text in self.__dict.suggest(self.__preedit_string):
                    self.__lookup_table.append_candidate(IBus.Text.new_from_string(text))
        text = IBus.Text.new_from_string(self.__preedit_string)
        text.set_attributes(attrs)
        self.update_auxiliary_text(text, preedit_len > 0)

        attrs.append(IBus.Attribute.new(IBus.AttrType.UNDERLINE,
                IBus.AttrUnderline.SINGLE, 0, preedit_len))
        text = IBus.Text.new_from_string(self.__preedit_string)
        text.set_attributes(attrs)
        self.update_preedit_text(text, preedit_len, preedit_len > 0)
        self.__update_lookup_table()
        self.__is_invalidate = False

    def __update_lookup_table(self):
        visible = self.__lookup_table.get_number_of_candidates() > 0
        self.update_lookup_table(self.__lookup_table, visible)


    def do_focus_in(self):
        print("focus_in")
        self.register_properties(self.__prop_list)

    def do_focus_out(self):
        print("focus_out")

    def do_reset(self):
        print("reset")

    def do_property_activate(self, prop_name):
        print("PropertyActivate(%s)" % prop_name)

